bm-timeline/src/Adoption.hs

```diff
@@ -0,0 +1,222 @@
+mknodes n = [0 .. n-1]
+
+processCsv :: FilePath -> [NodeId] -> IO ()
+processCsv basep nodes = do
+  -- let snodes = map (T.pack . show) nodes
+  parsed <- forM nodes $ loadNodeCsv basep
+  let hm_adopted0 = HM.empty :: HMAdopted
+  let hm_chain0 = HM.empty :: HMChain
+  let hm_leader0 = HM.empty :: HMLeader
+  let (hm_chain, hm_adopted, hm_leader) = collect parsed (hm_chain0,hm_adopted0,hm_leader0)
+  (tadopted0, tdiffusion0) <- reconstruct nodes (sort $ HM.keys hm_leader) (hm_chain,hm_adopted,hm_leader) [] []
+  -- sort by slot number; filter out entries with invalid slot number
+  let tadopted = sortOn (\(sn, _, _, _, _, _) -> sn) $ filter (\(sn, _, _, _, _, _) -> sn > 0) tadopted0
+  let tdiffusion = sortOn (\(sn, _, _, _, _, _) -> sn) tdiffusion0
+  -- calculate time between blocks (adopted)
+  let tbetween n = calculate_time_between hm_adopted n
+  let cdf_tbetween = foldl (\acc n -> pairs_in_columns(calc_cdf (tbetween n)) : acc) [] $ reverse nodes
+  let cdf_tb_headers = foldl (\acc n -> ["time between blocks (node " <> (T.pack . show) n <> ")","fraction node " <> (T.pack . show) n] : acc)
+                             [] $ reverse nodes
+  -- compute boxplot of block adoption time; prepend with source node id
+  let bp_adopted1 n = [textify n] <> (map textify $ calc_boxplot $ map (\(_, _, _, t, _, _) -> t) $ filter (\(_, _, src, _, _, _) -> src == n) tadopted)
+  let bp_adopted = foldl (\acc n -> bp_adopted1 n : acc) [] $ reverse nodes
+  -- compute boxplot of block diffusion time; prepend with source and target node ids
+  let bp_diffusion2 n1 n2 = [textify n1, textify n2] <> (map textify $ calc_boxplot $ map (\(_, _, _, t, _, _) -> t) $ filter (\(_, src, tgt, _, _, _) -> src == n1 && tgt == n2) tdiffusion)
+  let bp_diffusion1 n = foldl (\acc n2 -> bp_diffusion2 n n2 : acc) [] $ reverse nodes
+  let bp_diffusion = foldl (\acc n -> [["src node", "target node", "min","q1","median","q3","max"]] ++ bp_diffusion1 n ++ acc) [] $ reverse nodes
+  -- let bp_diff_headers = map (const ["src node", "tgt node", "min","q1","median","q3","max"]) nodes
+  withFile "adoption.csv" WriteMode $ \h ->
+      output_csv h $ named_columns $ [ (["slot", "timestamp", "node id", "tadopted", "num tx", "block size"], tuple6_in_columns(tadopted))
+                                     , (["node id", "min", "q1", "median", "q3", "max"], bp_adopted) ]
+                                    <> zip cdf_tb_headers cdf_tbetween
+  withFile "diffusion.csv" WriteMode $ \h ->
+      output_csv h $ named_columns [ (["slot", "src node", "tgt node", "tdiffusion", "num tx", "block size"], tuple6_in_columns(tdiffusion))
+                                   , (head bp_diffusion, tail bp_diffusion) ]
+                                    -- <> zip bp_diff_headers bp_diffusion
+
+  return ()
+
+
+calculate_time_between :: HMAdopted -> NodeId -> [Milliseconds]
+calculate_time_between hm n =
+  let adls = sortOn fst $ filter (\(_,(node,_,_,_)) -> node == n) $ HM.toList hm
```

Comment:
If we already have `where`-section in this function - do we need `let`-extression as well? Maybe move `adls` in `where`-section?

Reply:
yes, in general that's a good idea. Here "timedifference" is a local function, whereas "adls" is the preperation of the list that is used in the calculation. It just reads naturally if the "let" binding is preceding where it is used, so I keep the order in this case.